src/s_inter.c

```diff
@@ -138,6 +203,514 @@ void sys_set_temppath(void);
+static int rbsend_init();
+struct _rbskt { // t_rbskt
+    ring_buffer* rs_rb;
+    int rs_preserve_boundaries;
+    int rs_errno; // used by streaming sockets to report errors
+    int rs_closed; // used by streaming sockets to report the other end has gracefully closed the connection
+};
+
+static t_rbskt* rbskt_new(int preserve_boundaries) {
+    t_rbskt* x = (t_rbskt*)getbytes(sizeof(*x));
+    x->rs_rb = rb_create(RB_SIZE);
+    x->rs_preserve_boundaries = preserve_boundaries;
+    x->rs_errno = 0;
+    x->rs_closed = 0;
+    return x;
+}
+
+static void rbskt_free(t_rbskt* x) {
+    if(x) {
+        rb_free(x->rs_rb);
+        freebytes(x, sizeof(*x));
+    }
+}
+
+static int rbskt_ready(t_rbskt* x) {
+    return x && x->rs_rb && (x->rs_closed || x->rs_errno || rb_available_to_read(x->rs_rb));
+}
+
+int rbskt_bytes_available(t_rbskt* rbskt) {
+    return rb_available_to_read(rbskt->rs_rb);
+}
+
+// to ensure the correct order is preserved when calling send()/sendto()
+// directly from outside rb_dosendone(), always flush the corresponding
+// ringbuffer first
+// should_lock should be set to true if this is not called from the IO thread
+static void sendrb_flush(ring_buffer* rb, int should_lock) {
+    if(should_lock)
+        sys_lockio();
+    rb_dosend(rb, -1);
+    if(should_lock)
+        sys_unlockio();
+}
+
+// A sendto() that writes to a ring_buffer and is real-time safe
+static ssize_t rb_sendto(ring_buffer* rb, int socket, const void *buffer, size_t length, int flags, void* addr, size_t addrlen)
+{
+    //printf("rb_sendto: fd: %d, rb: %p \"%s\"\n", socket, rb, (char*)buffer);
+    size_t maxLength = rb_available_to_write(rb);
+    size_t desiredLength = sizeof(size_t) + length;
+    size_t actualLength = desiredLength <= maxLength ? desiredLength : maxLength;
+    if(actualLength < desiredLength)
+    {
+        error("rb buffer is full, sending packet from the audio thread");
+        sendrb_flush(rb, 1);
+        // then send out the new data
+        return sendto(socket, buffer, length, 0, addr, addrlen);
+    }
+    rb_write_to_buffer(rb, 5,
+            (char*)&socket, sizeof(socket),
+            (char*)&length, sizeof(length),
+            buffer, length,
+            (char*)&addrlen, sizeof(addrlen),
+            addr, addrlen
+        );
+    return actualLength;
+}
+
+// A send() that writes to a ring_buffer and is real-time safe
+ssize_t rb_send(ring_buffer* rb, int socket, const void *buffer, size_t length, int flags)
+{
+    return rb_sendto(rb, socket, buffer, length, flags, NULL, 0);
+}
+
+// a drop-in replacement for sendto(), exposed for public use, safe to call from a real-time thread
+ssize_t sys_sendto(int sockfd, const void *buf, size_t len, int flags, void* addr, size_t addrlen)
+{
+    if(rbsend_init())
+        return -1;
+    return rb_sendto(pd_this->pd_inter->i_rbsend, sockfd, buf, len, flags, addr, addrlen);
+}
+
+static int iothreadbuf_resize(int newsize)
```

Comment:
I think `iothreadbuf_resize` might not be necessary at all. Just allocate 65536 bytes and be done. While with gradual resizing you might save some memory, 65 kB is peanuts on todays systems.

There's actually one problem with resizing I forgot to mention: you can't know the packet size prior calling `recvfrom` (well, you can, but it requires another socket function call). If the buffer has 65536 bytes, it is guaranteed to hold the largest possible UDP packet. For TCP you don't ever need a higher buffer size anyway. See also https://github.com/pure-data/pure-data/pull/1122.

Needless to say that it will also make the code a lot simpler. Sorry for putting you on the wrong track here... :-/

Reply:
OK no problem

Reply:
09841966 
It was my fault for trying to be too clever about it.

Reply:
Actually, in the previous code I was discarding parts of the larger TCP messages (those that would exceed 65536).

This should fix it, can you have a look? ce80d1c0